docs/proposal/20241003-data-disk-support.md

```diff
@@ -0,0 +1,170 @@
+
+Creating new controllers is out of scope for the first pass of this enhancement.  The new disk will use the same controller as the primary disks.  The unit number for the disks on that controller will be in the order in which the disks are defined in the machine spec configuration.
+
+#### For User Story 2
+
+CAPV will treat all compute machine creation the same as the control plane machines.
+
+### API Design
+
+The following new struct has been created to represent all configuration for a vSphere disk
+
+```go
+// VSphereDisk describes additional disks for vSphere to be added to VM that are not part of the VM OVA template.
+type VSphereDisk struct {
+	// SizeGiB is the size of the disk (in GiB).
+	// +kubebuilder:validation:Required
+	SizeGiB int32 `json:"sizeGiB"`
+}
+```
+
+The above type has been added to the machine spec section
+
+```go
+// VirtualMachineCloneSpec is information used to clone a virtual machine.
+type VirtualMachineCloneSpec struct {
+    ...
+    // DataDisks holds information for additional disks to add to the VM that are not part of the VM's OVA template.
+    // +optional
+    DataDisks []VSphereDisk `json:"dataDisks,omitempty"`
+}
+```
+
+### Implementation Details
+
+The cloning process will now be able to add data disks to a VM.  Using the config provided in the VSphereCloneTemplate, the defined disks will be added to the virtual machine.  Each disk is added to the controller used by the primary disk.  Currently, there are no plans to allow the user to define a new controller (scsi/ide/etc) for these disks. 
+
+An example of what the VSphereMachineTemplate looks like when data disks are desired:
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: VSphereMachineTemplate 
+metadata:
+  name: test-machine-template
+  namespace: openshift-cluster-api
+spec:
+  template:
+    spec: 
+      template: CAPV_2_Disk
+      server: vcenter.example.com
+      diskGiB: 128
+      dataDisks:
+      - sizeGiB: 50
+      - sizeGiB: 90
+      cloneMode: linkedClone
+      datacenter: cidatacenter
+      datastore: /cidatacenter/datastore/vsanDatastore
+      folder: /cidatacenter/vm/multi-disk-k96l6
+      resourcePool: /cidatacenter/host/cicluster
+      numCPUs: 4
+      memoryMiB: 16384
+      network:
+        devices:
+        - dhcp4: true
+          networkName: ci-vlan-1240
+```
+
+In the above example, two data disks will be created during the clone process and placed after the OS disk.  The example shows a 50 GiB and a 90 GiB disk being configured and added.
+
+For each `dataDisks` definition, the clone procedure will attempt to generate a device VirtualDeviceConfigSpec that will be used to create the new disk device.  Each disk will be attached in the order in which they are defined in the template.  All disks defined in the vSphere OVA template will come first with the new disks being attached after.
```

Comment:
in the spec the user can only pass a size for a disk.
how can the user know which disk is which if the sizes are the same?

does the govmomi API types support names / indices that can be set from the source CAPV type optionally, so that the user can better tract what's going on in logs and API calls?

Reply:
I did not see any way to tag or name a disk from govmomi.  I can look a little more at govc to see what it also does.  I originally had a deviceName field so mentally you can put a name on the disk; however, here internally a lot of people here were hesitant on that idea.  If we can put a name on it, then that would be ideal for sure.

Also for the case where there are multiple disks with the same size, i think one the admin goes to create the machine config or whatever technique they are gonna use to assign the disk to a location, they can refer the disk to its DiskBySequence if preferred.  Depending on OS, these disks can also get mounted in Linux in a way that could potentially change between reboots.  RHEL did that for me, so mounting in my tests I did by `/dev/disk/by-diskseq`.  We are still looking into potentially making a boostrap-controller to handle the mounting / assigning of the disk locations.

Reply:
yes, just a Name or an Index field that can better help with tracking.
i'm sure users can track disks with the same size, but having the Name / Index would just make it easier.
so hopefully it's possible somehow.

Reply:
Ok, i've added a name field to the vsphere disk type.  let me know your thoughts.